coverage.c

```diff
@@ -0,0 +1,590 @@
+#include <math.h>    // round
+#include <stdbool.h>
+#include <string.h>
+#include <unistd.h>
+#include <sys/ioctl.h>
+
+#include "htslib/sam.h"
+#include "htslib/hts.h"
+#include "samtools.h"
+#include "sam_opts.h"
+
+const char *VERSION = "0.1";
+
+typedef struct {  // auxiliary data structure to hold a BAM file
+    samFile *fp;     // file handle
+    bam_hdr_t *hdr;  // file header
+    hts_itr_t *iter; // iterator to a region - NULL for us by default
+    int min_mapQ;    // mapQ filter
+    int min_len;     // length filter
+    unsigned int n_reads;  // records the number of reads seen in file
+    unsigned int n_selected_reads; // records the number of reads passing filter
+    unsigned long summed_mapQ; // summed mapQ of all reads passing filter
+} bam_aux_t;
+
+typedef struct {  // auxiliary data structure to hold stats on coverage
+    unsigned long long n_covered_bases;
+    unsigned long long summed_coverage;
+    unsigned long long summed_baseQ;
+    unsigned long long summed_mapQ;
+    unsigned int n_reads;
+    unsigned int n_selected_reads;
+    int32_t tid;    // chromosome ID, defined by header
+    int beg;
+    int end;
+    int bin_width;
+} stats_aux_t;
+
+int read_file_list(const char *file_list, int *n, char **argv[]);
+
+static int usage(int status) {
+    fprintf(stderr, "\nUsage: samtools coverage [options] in1.bam [in2.bam [...]]\n"
+                    "Options:\n"
+                    "  -o FILE           where to write output to [stdout]\n"
+                    "  -l <int>          read length threshold - ignore reads shorter than <int> [0]\n"
+                    "  -q <int>          base quality threshold [0]\n"
+                    "  -Q <int>          mapping quality threshold [0]\n"
+                    "  -H                print a file header\n"
+                    "  -m                show histogram instead of tabular output\n"
+                    "  -w <int>          number of bins in histogram. Set to 0 to use terminal width [50]\n"
+                    "  -r <chr:from-to>  Show region on chromosome chr. \n"
+                    "  -h                help (this page)\n"
+                    "  -v                version of this command\n"
+                    "\nGlobal options:\n");
+
+    sam_global_opt_help(stderr, "-.--.-");
+
+    fprintf(stderr, 
+            "\nThe tabular output is a simple tab-seperate table with 7 columns:\n"
+                    "  1. Reference name / chromosome\n"
+                    "  2. Start position\n"
+                    "  3. End position (or sequence length)\n"
+                    "  4. Number reads aligned to the region\n"
+                    "  5. Number of covered bases with depth >= 1\n"
+                    "  6. Percent of covered bases\n"
+                    "  7. Average coverage depth\n"
+                    "  8. Average baseQ in covered region\n"
+                    "  9. Average mapQ of selected reads\n"
+                    );
```

Comment:
I'd recommend outputting a one-line TSV header.
```
rname startpos endpos numreads numcov percentcov meandepth meanbaseq meanmapq
```

Reply:
Should the header be outputted by default? I'm for it, but I think most tools currently don't.  `samtools view` uses the `-h` flag to show the header, but also the SAM header is more involved than a simple TSV header. 

Other question: What should I change the percent-covered column name when reporting a number between 0 and 1? `proportioncov`? Can't think of a good name

Reply:
> Should the header be outputted by default? I'm for it, but I think most tools currently don't. samtools view uses the -h flag to show the header, but also the SAM header is more involved than a simple TSV header.

I would output the header by default. An option to disable the header would be useful (perhaps `--no-header`). `samtools coverage | tail +n2` works as well.

> Other question: What should I change the percent-covered column name when reporting a number between 0 and 1? proportioncov? Can't think of a good name

I'd suggest simply `coverage`. To me, genome coverage is the proportion of the genome that is covered, whereas depth of coverage is, well, the depth. =)
I'd suggest changing `numcov` to `coveredbases` or perhaps just `covbases`.

Reply:
Outputted the header by deault. Also changed the column names as suggested